### DISCUSSION TITLE
关于Rust中的类型转换的一个小问题 (discussion试试水

### DISCUSSION THREAD

**Discussion by 0130w** [LearningOS/rust-based-os-comp2022#175] [General]
我注意到在Rust Course开源书中的类型转换一节有下面这样的一段话

> "就算e as U1 as U2"是合法的，也不能说明"e as U2"是合法的

这段话大家是否有什么好的例子来理解呢？我暂时没有想到如何构造这样的一个前面合理而后面不合理的情况，我只能暂时想到说

```rust
fn main() {
    let a: i32 = 300;
    let b: i16 = (a as u8) as i16;
    println!("b = {}", b);
    let c: i16 = a as i16;
    println!("c = {}", c);
}
```

这两种转换方式最终损失的数据是不一样的，因此跑出来`b = 44`，`c = 300`.

**Comment by creatoy** [comment]
这句应该就是表达的 as 类型转换虽然可以进行级联的转换，但是不能认为实现了 e as U1 和 U1 as U2 就可以 e as U2 了。
我能找到的一个例子是枚举转整型再转裸指针的情况：
```rust
enum MyEnum {
    Zero,
}

fn main() {
    let _raw_ptr1 = MyEnum::Zero as i32 as *const i32;
    // let _raw_ptr2 = MyEnum::Zero as *const i32;
}
```
> 第二行就是不合法的

**Reply by 0130w** [reply]
感谢，非常有趣的例子!:D
